File: src/core/Systems/World.cs
using core.configs;
using core.Factories;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Timers;

namespace core.Systems {
	public class World {
		private static float worldDivision = 6, maxY = 8000, minY = -2000;
		private static Timer worldTimer = new Timer() { AutoReset = true, Enabled = false, Interval = 60000 };
		private static List<(float, float)> minMaxGroups = new List<(float, float)>();
		public static int hour = DefaultConfig.BOOTUP_HOUR;
		public static int minute = DefaultConfig.BOOTUP_MINUTE;

		public static void init() {
			generateGrid(worldDivision);
			worldTimer.Start();
		}
		public static void generateGrid(float division) {
			var total = maxY + Math.Abs(minY);
			for (int i = 0; i < division; i++) {
				minMaxGroups.Add((maxY - 2000 - (total / division) * i, maxY - (total / division) * i));
			}
		}
		public static void updateWorldTime() {
			minute += DefaultConfig.MINUTES_PER_MINUTE;
			if (minute >= 60) {
				minute = 0;
				hour += 1;
				var lastIndex = DefaultConfig.WEATHER_ROTATION.Count - 1;
				var endElement = DefaultConfig.WEATHER_ROTATION[lastIndex];
				DefaultConfig.WEATHER_ROTATION.RemoveAt(lastIndex);
				DefaultConfig.WEATHER_ROTATION.Insert(0, endElement);
			}
			if (hour >= 24) hour = 0;
		}
		public static int getGridSpace(MyPlayer player) {
			var gridSpace = minMaxGroups.FindIndex(data => player != null && player.Exists && player.Position.Y > data.Item1 && player.Position.Y < data.Item2);
			return Math.Max(0, gridSpace);
		}
		public static string getWeatherByGrid(int gridSpace) {
			return DefaultConfig.WEATHER_ROTATION[gridSpace];
		}
	}
}

File: src/core/server/systems/world.ts
import * as alt from 'alt-server';
import { DefaultConfig } from '../configs/settings';

const worldDivision = 6;
const maxY = 8000;
const minY = -4000;

export class World {
    static minMaxGroups: Array<{ minY: number; maxY: number }>;
    static hour: number = DefaultConfig.BOOTUP_HOUR;
    static minute: number = DefaultConfig.BOOTUP_MINUTE;

    static generateGrid(division: number): void {
        let groups: Array<{ minY: number; maxY: number }> = [];
        let total = maxY + Math.abs(minY);
        for (let i = 0; i < division; i++) {
            const result = { maxY: maxY - (total / division) * i, minY: maxY - 2000 - (total / division) * i };
            groups.push(result);
        }
        World.minMaxGroups = groups;
    }

    static updateWorldTime(): void {
        World.minute += DefaultConfig.MINUTES_PER_MINUTE;
        if (World.minute >= 60) {
            World.minute = 0;
            World.hour += 1;
            const endElement = DefaultConfig.WEATHER_ROTATION.pop();
            DefaultConfig.WEATHER_ROTATION.unshift(endElement);
        }
        if (World.hour >= 24) World.hour = 0;
    }

    static getGridSpace(player: alt.Player): number {
        const gridSpace = World.minMaxGroups.findIndex((pos) => player && player.valid && player.pos.y > pos.minY && player.pos.y < pos.maxY);
        return gridSpace === -1 ? 0 : gridSpace;
    }

    static getWeatherByGrid(gridIndex: number): string {
        return DefaultConfig.WEATHER_ROTATION[gridIndex];
    }
}

alt.setInterval(World.updateWorldTime, 60000);
World.generateGrid(worldDivision);
